### PR TITLE
Support Japanese punctuation in path

### PR DESCRIPTION
**Pull request by adamdavidcole**
Fix for issue https://github.com/gregjacobs/Autolinker.js/issues/297: full width brackets in url not work correctly

This PR adds support for Japanese punctuation marks (and CJK punctuation in general).

Previously links that looked like this would fail at the first punctuation mark: https://ja.wikipedia.org/wiki/Yahoo!#ドットコムバブル（2000年_-_2001年）

The regex used contains most Japanese common characters which is a superset of all punctuation marks and there is likely some overlap with the `alphaCharsAndMarksStr` regex. We could pull out all the individual punctuation characters from the Japanese unicode blocks included, but we would likely miss some. The overlap seems 🆗 to me, but open to suggestions. 

Something to consider is getting a regex for the unicode category `\p{Punctuation}` to support even more languages -- but that would be a riskier change. This smaller addition should be a more predictable change. 

### Demo
Screen Shot 2020-02-21 at 5 24 24 PM

### CODE REVIEW

**Comment by adamdavidcole** [comment]
Thought about this a bit more -- think we need to more accurately target common Japanese punctuation marks to avoid including natural link delimiters like the Japanese full width space character: "　".

**Comment by adamdavidcole** [comment]
Alright: gave this a second, much more targeted attempt. 

This add support for the common Japanese punctuation defined here: https://en.wikipedia.org/wiki/Japanese_punctuation

I also made a slight refactor for how `urlSuffixRegex` is built by pulling out the additional punctuation into its own variable. Let me know if you think it's easier to read. Also, please double check that I correctly moved what was in the larger regex to the smaller parts (aka. `latinIntraSuffixChars` and `latinEndOfSuffixChars`).

I am going to do a little more research if certain punctuation like 。should ever be in a URL, but I think this is a step in the right direction. 

### Before
Screen Shot 2020-02-24 at 6 33 35 PM

### After
Screen Shot 2020-02-24 at 6 35 17 PM

File: src/matcher/url-matcher.ts [hunk 1]
```diff
@@ -74,9 +74,19 @@ export class UrlMatcher extends Matcher {
...
-		    urlSuffixRegex = new RegExp( '[/?#](?:[' + alphaNumericAndMarksCharsStr + '\\-+&@#/%=~_()|\'$*\\[\\]{}?!:,.;^\u2713]*[' + alphaNumericAndMarksCharsStr + '\\-+&@#/%=~_()|\'$*\\[\\]{}\u2713])?' );
+			// special characters that are permitted within the URL suffix, but not as the final character
+			latinIntraSuffixChars = /\-\+&@#\/%=~_\(\)\|\'\$\*\[\]{}\?!:,\.;\^\u2713/.source,
+			// Japanese special characters based on common punctuation defined here: https://en.wikipedia.org/wiki/Japanese_punctuation
+			// in addition to the full width variants of the latin special chars
+			japaneseIntraSuffixChars = /ー＋＆＠＃％＝〜＿｜＄￥＊｛｝（）［］【】「」『』〝〟？！：、，…	‥。・＾｡･､/.source,
+
+			// special characters that are permitted at end of URL suffix
+			latinEndOfSuffixChars = /\-\+&@#\/%=~_\(\)\|\'\$\*\[\]{}\u2713/.source,
+			japaneseEndOfSuffixChars = /ー＋＆＠＃％＝〜＿｜＄￥｛｝（）［］【】「」『』〝〟/.source,
```
Comment (gregjacobs) [inline src/matcher/url-matcher.ts]: Liking the breakout into these extra variables!

Kind of minor, but to remove duplication, could we build `latinIntraSuffixChars` and `japaneseIntraSuffixChars` out from `latinEndOfSuffixChars` and `japaneseEndOfSuffixChars`? Maybe something like the following?

```ts
latinIntraSuffixChars = latinEndOfSuffixChars + /\?!:,\.;\^/.source;
```

Comment (gregjacobs) [reply src/matcher/url-matcher.ts]: Good commenting too :)

File: tests/autolinker-url.spec.ts [hunk 2]
```diff
@@ -1002,6 +1002,52 @@ describe( "Autolinker Url Matching -", () => {
...
+
+			expect( result ).toBe( '<a href="https://ja.wikipedia.org/wiki/Yahoo!#ドットコムバブル（2000年_-_2001年）">ja.wikipedia.org/wiki/Yahoo!#ドットコムバブル（2000年_-_2001年）</a>' );
+
+		} );
+
+		it( 'should exclude concluding Japanese parenthesis if URL is wrapped in parens', () => {
+			var result = autolinker.link( "（https://ja.wikipedia.org/wiki/Yahoo!#ドットコムバブル2000年_-_2001年）" );
+
+			expect( result ).toBe( '（<a href="https://ja.wikipedia.org/wiki/Yahoo!#ドットコムバブル2000年_-_2001年">ja.wikipedia.org/wiki/Yahoo!#ドットコムバブル2000年_-_2001年</a>）' );
+		} );
```
Comment (gregjacobs) [inline tests/autolinker-url.spec.ts]: Nice job on these tests!

**Comment by gregjacobs** [comment]
Good research on this btw, and thanks for doing that. I'm sure it was no short task!

**Comment by adamdavidcole** [comment]
Appreciate the commnets! I think I'm gonna need to pause on this one for a bit. 

The complication with Japanese is that they don't use spaces between characters -- if we add additional punctuation as valid path characters it could become even more difficult to delimit the end of a link in Japanese. 

I'll revisit this in a bit, maybe it's as simple as further limiting which additional Japanese characters to allow (for example, excluding Japanese periods and commas).

**Comment by gregjacobs** [comment]
Gotcha, kk. Let me know what you find!

**Pull request closed without merging**